.github/workflows/test_functional.yml

```diff
@@ -106,54 +133,79 @@ jobs:
           ssh -vv _remote_background_indep_poll hostname
 
       - name: Test
+        id: test
+        timeout-minutes: 35
+        continue-on-error: true
         run: |
-          # Use non-UTC time zone
-          export TZ=XXX-05:30
-
-          if [[ "${BASE}" == 'tests/k' ]]; then
+          if [[ '${{ matrix.test-base }}' == 'tests/k' ]]; then
               NPROC=4
           else
               NPROC=8
           fi
-          # NOTE: $BASE is purposefully un-quoted
+          # NOTE: test base is purposefully un-quoted
           etc/bin/run-functional-tests \
               -j "${NPROC}" \
-              --state=save ${BASE} \
-          || (
-              echo '=== Re-running Failed Tests ==='
-              CHUNK= etc/bin/run-functional-tests \
-                  -j "${NPROC}" \
-                  --state=save,failed
-          )
+              --state=save \
+              ${{ matrix.test-base }} \
+              || (touch tests-did-not-time-out && false)
+
+      - name: Re-run failed tests
+        timeout-minutes: 10
+        if: steps.test.outcome == 'failure'
+        run: |
+          if [[ ! -f tests-did-not-time-out ]]; then
+            echo "the preious test run timed out -> not re-running tests"
+            # this false causes the workflow to fail
+            false
+          fi
+          # TODO: make the tests deterministic so we don't need to do this
+          etc/bin/run-functional-tests \
+              -j 1 \
+              -v \
+              --state=save,failed ${{ matrix.test-base }}
+
+      - name: Copy cylc-run out of container
+        if: failure() && steps.test.outcome == 'failure' && contains(matrix.platform, '_remote')
+        run: |
+          # pick the first host in the list
+          host="$(cut -d ' ' -f 1 <<< "${{ matrix.platform }}")"
+          # copy back the remote cylc-run dir
+          rsync -av \
+            "${host}:/root/cylc-run/" \
+            "${HOME}/cylc-run/${host}/"
```

Comment:
This means that you should expect to find the remote container's `cylc-run` dir in the uploaded artefact.